File: LeetCode/Medium/0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        int l = 0, r = nums.length - 1, mid = 0;
        int lVal = nums.length, rVal = -1;
        
        while (l <= r){
            mid = (l + r) / 2;
            
            if (nums[mid] == target) lVal = Math.min(lVal, mid);

            if (nums[mid] >= target) r = mid - 1;
            else l = mid + 1;
        }

        l = lVal;
        r = nums.length - 1;

        while (l <= r){
            mid = (l + r) / 2;
            
            if (nums[mid] == target) rVal = Math.max(rVal, mid);

            if (nums[mid] <= target) l = mid + 1;
            else r = mid - 1;
        }

        if (lVal == nums.length) lVal = -1;
        if (rVal == -1) rVal = -1;

        int[] ret = {lVal, rVal};

        return ret;
    }
}

File: LeetCode/Medium/0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.go
func searchRange(nums []int, target int) []int {
    l, r, lVal, rVal, mid := 0, len(nums)-1, len(nums), -1, 0
    
    for l <= r {
        mid = (l + r) / 2
        
        if nums[mid] == target { lVal = min(lVal, mid) }
        
        if nums[mid] >= target {
            r = mid - 1
        } else { l = mid + 1 }
    }
    
    if lVal == len(nums) { return []int{-1, -1} }
    
    l = lVal
    r = len(nums) - 1
    
    for l <= r {
        mid = (l + r) / 2
        
        if nums[mid] == target { rVal = max(rVal, mid) }
        
        if nums[mid] <= target {
            l = mid + 1
        } else { r = mid - 1 }
    }
    
    return []int{lVal, rVal}
}

func min(a, b int) int {
    if a > b { return b }
    return a
}

func max(a, b int) int {
    if a > b { return a }
    return b
}